### ISSUE TITLE
PAR-2 — Nightly property/fuzz parity suite and the frozen fp32 bit-reproduction golden

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1573]
**Depends on:** PAR-1 (#1572) · **Blocks:** —

**Context:** PAR-1 (#1572) makes the frozen legacy `mace/` a live in-process differential oracle, but it runs **tiny fixed fixtures per PR** — enough to catch structural bugs, blind to parity drift that only shows on *weak inputs* (degenerate cells, empty neighbourhoods, unusual `lmax`, dtype mixes). PAR-2 (#1573) adds the heavy half of the fast-vs-heavy split: a **property-based (hypothesis) fuzz-parity suite running in `nightly.yaml`**, so subtle legacy-vs-v1 drift on adversarial inputs cannot pass silently, at zero per-PR CI cost. It also closes a precision blind spot no fp64 CPU test can see: `scatter_add_` on CUDA is **order-nondeterministic**, so ~1e-4 fp32 differences hide under fp64 tolerances (legacy's edge→node reductions go through the vendored `mace/tools/scatter.py`). The countermeasure is a **frozen fp32 bit-reproduction golden** captured with a deterministic `segment_reduce`, gated on **max-abs-error at fp32 machine epsilon** — never a loose tolerance. This same golden is what FM-2 (#1587) later gates converted foundation-model artifacts on, so it must be committed and stable here.

**Interface & constraints:**

1. **The input classes are a closed list** (each is a named hypothesis strategy; each targets a known failure mode):

| Class | What it stresses |
|---|---|
| random triclinic cells / mixed per-axis PBC | cell handling; per-axis `pbc` flags are passed straight to the neighbour builder — mixed PBC is fully supported semantics, not an edge case |
| cell smaller than cutoff | **periodic self-images**: e.g. 1 atom in a 2 Å cell with a 3 Å cutoff yields 18 self-image edges, none with unit shift 0; self-edge removal drops only `sender==receiver ∧ unit_shifts==0` (legacy: `mace/data/neighborhood.py:72-81`) |
| aperiodic and zero-vacuum-slab geometries | the **three effective-cell regimes** DATA-3 (#1571) pins (`mace/data/neighborhood.py:51-60`): fully aperiodic returns the extended construction cell, any-periodic returns the physical one, a zero physical row on a non-periodic axis is patched. A regression here is invisible in energies and shows only in stress and k-space terms |
| empty-neighbourhood nodes | zero-edge nodes/systems are **legal inputs**; catches any op that derives `num_nodes` from `edge_index.max()` instead of the graph metadata |
| high `lmax` | exercises the dispatch layer near `capabilities().max_lmax`; strategies must respect the backend's declared capability, not skip silently past it |
| mixed fp32/fp64 | per-quantity dtype behaviour (fp64 E0/scale-shift alongside model-dtype features) under non-default dtype combinations |

Cases that vary model hyperparameters (high `lmax`) cannot reuse the committed tiny anchors: instantiate the legacy model with a seeded init and map its weights onto v1 through the in-process converter (FM-00 (#1565)), so both stacks run byte-identical parameters — the same discipline as PAR-1 (#1572)'s single-step gradient-parity test.

2. **Comparison rules.** Every fuzz case runs the same structure through frozen legacy and v1 using PAR-1 (#1572)'s harness — including its **global process-state snapshot/restore** (env vars, default dtype, legacy toolchain optimization defaults, torch/numpy RNG) or forked-subprocess isolation — and compares E/F/stress with the **edit-locked tolerance table** (fp64 CPU 1e-6 eV / 1e-6 eV/Å; fp32 1e-3 relative), tolerance row selected by the case's dtype. This ticket may not edit the tolerance table (a tolerance change is a separate, justified PR).

3. **Reproducibility of failures.** A nightly fuzz failure must be locally reproducible: persist the hypothesis example database (or the failing seed + serialized structure) as a CI artifact, and print the seed in the failure message. A red nightly that cannot be replayed is a non-deliverable. Note the artifact trap nightly has already paid for once: hypothesis's database is a **dotfile directory**, so the upload step needs `include-hidden-files` — the same omission that silently emptied the durations upload and now guards the `coverage-full` shards.

4. **The fp32 bit-reproduction golden.** Captured once on a fixed fixture with the reference backend's **deterministic `segment_reduce`** (fixed accumulation order — matching legacy CPU accumulation order is a PAR-1 (#1572) conformance obligation this ticket relies on, not redefines), stored frozen under the golden references. The check is **bitwise-tight**: max-abs-error at fp32 machine epsilon. Golden discipline applies — it is never regenerated inside this or any feature ticket.

5. **Per-input-class coverage counters.** Each nightly run records how many examples of each class in the table above were actually generated and executed, and **fails if any class count is zero** — hypothesis strategies can silently shrink or filter a class out of existence, and a fuzz suite that no longer produces self-images is a green lie.

6. **Placement.** The suite lives in `tests/parity/` (the dual-import allowlist module — the only place, with `mace_launcher`, allowed to import both stacks) and is wired into `nightly.yaml` as its own job, not into the per-PR set. It must not be folded into `coverage-full`: that job is already the only one installing every extra at once on a 16 GB runner, which is why it runs as 3 shards plus a `coverage-report` combine step.

**Task:**
1. Write hypothesis strategies in `tests/parity/` for the six input classes: triclinic cell + per-axis PBC generator, sub-cutoff cells (self-images), aperiodic/zero-vacuum-slab geometries exercising the three cell regimes, isolated/zero-edge structures, `lmax` sweep bounded by `capabilities().max_lmax`, and fp32/fp64 dtype mixes.
2. Build the property test: each generated case runs legacy and v1 through PAR-1 (#1572)'s isolation harness and asserts E/F/stress parity at the per-dtype tolerance row.
3. Capture and commit the frozen fp32 bit-reproduction golden (deterministic `segment_reduce`, fixed fixture, fixed seed) with a sidecar recording the exact generation command; add the test asserting max-abs-error ≤ fp32 machine epsilon against it.
4. Implement the per-input-class coverage counters; emit them in the test summary and fail on any zero-count class.
5. Wire the suite into `.github/workflows/nightly.yaml` as its own job (nightly-only marker; per-PR runtime unchanged) and upload the hypothesis example database / failing seeds as a job artifact **with `include-hidden-files`**.

**Out of scope:** GPU execution of the fuzz suite (accelerated-backend parity incl. double-backward is BKD-3 (#1567); fleet gating lives in the GPU CI); any edit to the tolerance table; regenerating existing goldens; FM-2 (#1587)'s consumption of the fp32 golden for converted foundation artifacts.

**Acceptance criteria:**
- [ ] The fuzz-parity suite is green in `nightly.yaml` with every input class in the table exercised (all coverage counters > 0), and does not run on per-PR CI.
- [ ] A deliberately injected v1 perturbation on a weak input (e.g. dropping self-image edges, or returning the physical cell for a fully aperiodic system) is caught by the suite, and the failure replays locally from the persisted seed/artifact.
- [ ] The fp32 bit-reproduction golden is committed with its generation sidecar; the deterministic-`segment_reduce` path reproduces it at fp32 machine epsilon (max-abs-error gate).
- [ ] A zero-count input class fails the nightly job (counter guard verified by disabling one strategy in a test).
- [ ] The hypothesis database artifact actually contains files (the hidden-file upload trap is covered by an assertion, not by hope).
- [ ] The edit-locked tolerance table is unmodified by this PR.

**Verify:**
```bash
python -m pytest tests/parity/test_fuzz_parity.py --hypothesis-seed=0 -v      # full nightly profile, fixed seed, CPU
python -m pytest tests/parity/test_fuzz_parity.py -k coverage_counters -v     # class-counter guard
python -m pytest tests/parity/test_fp32_bit_golden.py -v                      # machine-epsilon gate vs frozen golden
```
**Review focus:** the fp32 machine-epsilon gate — that the deterministic `segment_reduce` accumulation order is actually pinned (a loose tolerance here silently re-opens the fp32 nondeterminism blind spot) — and that the strategies genuinely generate each adversarial class rather than shrinking it away (coverage counters honest).
## Phase 3 — Training pipeline, config/CLI, electrostatics core, fine-tuning (17 tickets)

> The 68 black-box test functions (54 in `tests/workflows` + 14 in `tests/integrations`) are re-run under `--engine v1` for migrated capabilities. `tests/helpers.py` is edited **once** so `run_mace_train` invokes the launcher entry point rather than `mace/cli/run_train.py` directly — it currently sets `run_train = REPO_ROOT / "mace" / "cli" / "run_train.py"` (`tests/helpers.py:31`) and builds `cmd = [sys.executable, str(script)]` (`:201`), so the change is one indirection through `mace_launcher`, engine selected by env. CLI-1 (#1579) is sized to include a **faithful port of the legacy argparse flag surface** so the v1 CLI accepts the pinned flags identically. The electrostatics core lands in this phase, not with the rest of electrostatics: ELEC-1 (#1591) (reference solver + dispatch layer) and ELEC-2 (#1592) (`PolarMACE` on the new architecture) are consumed by DEP-1a (#1634)'s polar/LES calculator surfaces and by FT-4 (#1637), so GATE-3 (#1585) cannot close without them. ELEC-3 (#1593) and ELEC-4 (#1635) stay in Phase 5, and the MACE-Polar artifact conversion goes with the rest of the roster in FM-2 (#1587).
